### PR TITLE
chore(l1-contracts): Migrate L1 Contracts to Custom Errors

### PR DESCRIPTION
**Pull request by koloz193**
# What ❔

Moving all of the l1 contracts to use custom errors instead of string based reverts

pt 2 of https://github.com/matter-labs/era-contracts/pull/473 

Gas Report From Before: https://gist.github.com/koloz193/c8f39f6b61411e6b6a3ddebd8a8869df
Gas Report From After: https://gist.github.com/koloz193/beb78cb4ed2aa0919d7adda72b288a07

## Checklist

- [ ] PR title corresponds to the body of PR (we generate changelog entries from PRs).
- [ ] Tests for the changes have been added / updated.
- [ ] Documentation comments have been added / updated.

### CODE REVIEW

File: l1-contracts/contracts/common/L1ContractErrors.sol [hunk 1]
```diff
@@ -0,0 +1,63 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.24;
+
+error Unauthorized(address caller);
```
Comment (vladbochok) [inline l1-contracts/contracts/common/L1ContractErrors.sol:4]: Better to place errors in the interface version of the contract. So we will see which error supposed to be in which contract + there will be no ridiculously big imports

File: l1-contracts/contracts/bridge/L1SharedBridge.sol [hunk 2]
```diff
@@ -168,24 +206,32 @@ contract L1SharedBridge is IL1SharedBridge, ReentrancyGuard, Ownable2StepUpgrade
...
+                revert ValueMismatch(balanceAfter - balanceBefore, legacyBridgeBalance);
+            }
             chainBalance[_targetChainId][_token] = chainBalance[_targetChainId][_token] + legacyBridgeBalance;
         }
     }
 
     function receiveEth(uint256 _chainId) external payable {
-        require(BRIDGE_HUB.getHyperchain(_chainId) == msg.sender, "receiveEth not state transition");
+        if (BRIDGE_HUB.getHyperchain(_chainId) != msg.sender) {
+            revert RevertWithMsg(ERR_RECEIVE_ETH_BRIDGE);
```
Comment (vladbochok) [inline l1-contracts/contracts/bridge/L1SharedBridge.sol:233]: Why use message here?

File: l1-contracts/contracts/common/L1ContractErrors.sol (hunk 1 above)
Comment (koloz193) [reply l1-contracts/contracts/common/L1ContractErrors.sol:4]: we'll likely end up with the same number of imports unless they live within the interface (but then we need to duplicate them in the tests) and a single file lets us remove duplicates. I don't feel strongly though on this. 

Maybe @dnkolegov does?

File: l1-contracts/contracts/bridge/L1SharedBridge.sol (hunk 2 above)
Comment (koloz193) [reply l1-contracts/contracts/bridge/L1SharedBridge.sol:233]: it seemed to be a hyper-specific error but I can change it out

File: l1-contracts/contracts/upgrades/BaseZkSyncUpgrade.sol [hunk 3]
```diff
@@ -203,10 +206,9 @@ abstract contract BaseZkSyncUpgrade is ZkSyncHyperchainBase {
 
         // We want the hashes of l2 system upgrade transactions to be unique.
         // This is why we require that the `nonce` field is unique to each upgrade.
-        require(
-            _l2ProtocolUpgradeTx.nonce == _newProtocolVersion,
-            "The new protocol version should be included in the L2 system upgrade tx"
-        );
+        if (_l2ProtocolUpgradeTx.nonce != _newProtocolVersion) {
+            revert NewProtocolVersionNotInUpgradeTxn();
```
Comment (vladbochok) [inline l1-contracts/contracts/upgrades/BaseZkSyncUpgrade.sol:210]: What is Upgrde Tx**n**? Txn == transaction? Usually people use tx, but no strong opinion just for my understanding

File: l1-contracts/contracts/upgrades/BaseZkSyncUpgrade.sol [hunk 4]
```diff
@@ -221,38 +223,40 @@ abstract contract BaseZkSyncUpgrade is ZkSyncHyperchainBase {
...
+        }
         uint256 length = _factoryDeps.length;
 
         for (uint256 i = 0; i < length; ++i) {
-            require(
-                L2ContractHelper.hashL2Bytecode(_factoryDeps[i]) == bytes32(_expectedHashes[i]),
-                "Wrong factory dep hash"
-            );
+            if (L2ContractHelper.hashL2Bytecode(_factoryDeps[i]) != bytes32(_expectedHashes[i])) {
+                revert InvalidHash();
```
Comment (vladbochok) [inline l1-contracts/contracts/upgrades/BaseZkSyncUpgrade.sol:236]: Shall we add params here?

File: l1-contracts/contracts/upgrades/BaseZkSyncUpgrade.sol [hunk 5]
```diff
@@ -221,38 +223,40 @@ abstract contract BaseZkSyncUpgrade is ZkSyncHyperchainBase {
...
-        require(
-            _newProtocolVersion > previousProtocolVersion,
-            "New protocol version is not greater than the current one"
-        );
-        require(
-            _newProtocolVersion - previousProtocolVersion <= MAX_ALLOWED_PROTOCOL_VERSION_DELTA,
-            "Too big protocol version difference"
-        );
+        if (_newProtocolVersion <= previousProtocolVersion) {
+            revert InvalidProtocolVersion();
```
Comment (vladbochok) [inline l1-contracts/contracts/upgrades/BaseZkSyncUpgrade.sol:246]: `TooSmallProtocolVersion`

File: l1-contracts/contracts/upgrades/BaseZkSyncUpgrade.sol [hunk 6]
```diff
@@ -221,38 +223,40 @@ abstract contract BaseZkSyncUpgrade is ZkSyncHyperchainBase {
...
+        }
+        if (_newProtocolVersion - previousProtocolVersion > MAX_ALLOWED_PROTOCOL_VERSION_DELTA) {
+            revert InvalidProtocolVersion();
```
Comment (vladbochok) [inline l1-contracts/contracts/upgrades/BaseZkSyncUpgrade.sol:249]: `TooBigProtocolVersion`

**Review by vladbochok — CHANGES_REQUESTED** [review]

File: l1-contracts/contracts/common/L1ContractErrors.sol [hunk 7]
```diff
@@ -0,0 +1,233 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.24;
+
+// 0x8e4a23d6
```
[the remark this answers is not in the document]
Comment (koloz193) [reply l1-contracts/contracts/common/L1ContractErrors.sol:4]: you can always check by recalculating. its meant to help instead of be used directly
